Change to not expose unsupported aliases/cmdlets in Unix

iSazonov opened the pull request:
Fix #3577.

Make changes to not expose the aliases "gin", "gsv", "sasv" and "spsv" in Unix platforms.
Also refactored aliases related tests and added new tests covering the complete list of built-in aliases/cmdlets.

iSazonov committed:
Remove 'gin' alias on Unix

mirichmo commented on test/powershell/Modules/Microsoft.PowerShell.Management/Get-ComputerInfo.Tests.ps1:
```diff
@@ -1332,11 +1332,6 @@ try {
             }
         }
 
-        It "Verify that alias 'gin' exists" {
```
Rather than removing this test, can you skip it for non-windows platforms?

iSazonov replied:
The idea (from @lzybkr) is that we should put all alias tests in one place (Default-Aliases.Tests.ps1) rather than spread on many files.

daxian-dbw commented:
@iSazonov thanks for the fix!
There are a few other aliases that are not supposed to be exposed on Unix. Do you think they can be taken care of in the same PR? (absolutely not required 😄)
```
PS /> Get-Alias | ? { gcm $_.Definition -ea SilentlyContinue > $null; !$? }                         

CommandType     Name                                               Version    Source               
-----------     ----                                               -------    ------               
Alias           gin -> Get-ComputerInfo                            3.1.0.0    Microsoft.PowerShe...
Alias           gsv -> Get-Service                                                                 
Alias           sasv -> Start-Service                                                              
Alias           spsv -> Stop-Service           
```

iSazonov committed:
Fix aliases
Fix 'gsv','sasv' and 'spsv'
Fix 'fhx' alias
Fix 'cfs' alias
Fix 'gcb' and 'scb' aliases
Fix 'gtz' and 'stz' aliases
Fix 'gin'

iSazonov commented:
@daxian-dbw I am glad to fix this aliases too.

Now I catched interensting issue with "gin" alias. Please help me understand. As for `Get-Service` cmdlet I add the alias in `InitialSessionState.cs` and remove from .psd1 files. After that I start PowerShell, ipmo .\Build.psm1 and get error "AllScope option cannot be removed from the alias 'gin'". CI tests is failed too as you see. Then I remove `[Alias("gin")]` from `GetComputerInfoCommand.cs` (locally only) - now all work well.
So my question is what is right way? Move aliases to `InitialSessionState.cs` or add Alias attribute to cmdlets and exclude them on unsupported platform? Or Alias attribute is for third-party cmdlets only and we should remove it from `Get-ComputerInfo`?
And how `Get-ComputerInfo` is excluded from compiling code on Unix?

daxian-dbw commented:
> So my question is what is right way?

What I would do is:
1. For cmdlets that are not supposed to work on Unix, like `Get-ComputerInfo`, exclude the source code from unix build, by enclosing the code with in `#if !UNIX` and `#endif`. Please take `TimeZoneCommands.cs` as an example.
2. For aliases that declared in `InitialSessionState.cs`, move the declarations to the `#if !UNIX/endif` section. You will find the section in the part of code that declares all `SessionStateAliasEntry` instances.

For cmdlets like `*-Clipboard`, they have already been excluded from the build in powershell core. You can take a look at the .csproj file of the corresponding project.

Basically, there are also 4 aliases that are currently exposed on unix plats by mistake. You can run the following command to observe.
```
Get-Alias | ? { gcm $_.Definition -ea SilentlyContinue > $null; !$? }                         

CommandType     Name                                               Version    Source               
-----------     ----                                               -------    ------               
Alias           gin -> Get-ComputerInfo                            3.1.0.0    Microsoft.PowerShe...
Alias           gsv -> Get-Service                                                                 
Alias           sasv -> Start-Service                                                              
Alias           spsv -> Stop-Service         
```

iSazonov commented:
@daxian-dbw Thanks! I still don't understand

1. If currently  Get-ComputerInfo don't excluded by `#if !UNIX  and  #endif ` it is bug and we should have tests to control such problems?
2. What is the preferred way to add an alias in `InitialSessionState.cs` or by an attribute `[Alias("alias")]`? In last commits I used first way. We should use the single preferred way to avoid errors (which addressed in the PR). It seems the first way gives a more quick start, the second way is more intuitive.

@lzybkr Could you please comment too?

daxian-dbw commented:
> If currently Get-ComputerInfo don't excluded by #if !UNIX and #endif it is bug and we should have tests to control such problems?

Yes, we can have a test for it. Basically, we need to get the list of types that shouldn't be available on different platforms, for example, `[Microsoft.PowerShell.Commands.GetComputerInfoCommand]` should fail on Unix plats.

> What is the preferred way to add an alias in InitialSessionState.cs or by an attribute [Alias("alias")]? In last commits I used first way. We should use the single preferred way to avoid errors (which addressed in the PR). It seems the first way gives a more quick start, the second way is more intuitive.

Declaration of aliases in `InitialSessionState.cs` came from the days before `[Alias]` attribute was made supported for command types (previously `[Alias]` was only supported for parameters). So they are legacy code. For new cmdlet code, `AliasAttribute` should be used.

For the legacy code, I'm inclined to keep them that way, as I'm not sure if changing them would cause some unobvious regression. For new cmdlet implementation, I think `AliasAttribute` is the way to go.

iSazonov commented:
Can we create new attribute like `Platform("Unix")` to mark cmdlet classes? Can this approach make our tests easier?

What unobvious regression you expect with aliases? I believe that there is no functional code which we can break down, no new code. On the other hand, if we move the aliases in cmdlets (`[Alias]`) we get rid of some of the complexities that are trying to fix here.

daxian-dbw commented:
> What unobvious regression you expect with aliases?
Those alias declarations are publicly available, via `InitialSessionState.Commands`. For example:
```
PS> $iis = [initialsessionstate]::CreateDefault()
PS> $iis.Commands.Name | ? { $_ -eq 'gps' }
gps
```
So it's hard to anticipate how people will depend on them ... that's why I said _unobviously_ regression.

> Can we create new attribute like Platform("Unix") to mark cmdlet classes? Can this approach make our tests easier?

A class that should be excluded from a platform may not be a cmdlet, so I don't think a new attribute will help.

Not excluding a file that is supposed to be excluded on a platform won't be a big issue really. I think we can just fix it when it comes up.

iSazonov commented:
Clear about new attribute.

I tried to add `gin` alias to `InitialSessionState.cs` and use `Alias('gin')]` (the [code](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/InitialSessionState.cs#L5886) add cmdlet aliases in a session) - in both cases, I see the alias in a session state. Both options work equally. So it looks safe if we move the cmdlet aliases in cmdlets and exclude the files (or the aliases) with #if !UNIX  and  #endif if we need. The only public change will be (not breaking change) that we will remove aliases that should not be there (aliases for not ported cmdlets or aliases unwanted on Unix) and add the ones we want to appear. So users get right thinks with ` [initialsessionstate]::CreateDefault()`.

daxian-dbw commented:
@iSazonov thanks for the extra validation exercise! Then I'm fine moving the aliases from `InitialSessionState` to individual cmdlets. Do you want to do that in a separete PR or address all in this PR?

iSazonov commented:
@daxian-dbw Thanks for approve! I believe we can continue here to preserve the usefull discussion.

I have already started to prepare new test.

iSazonov committed:
Move aliases to InitialSessionState.cs

iSazonov commented:
I created new file with tests to check default alias and cmdlet lists. 
Alias tests is from `Default-Aliases.Tests.ps1`. Later we should remove the test from `Default-Aliases.Tests.ps1`.
I would use the session state to get default alias and cmdlet lists instead of `Get-Alias` and `Get-Command` but latest is more simple use.

@daxian-dbw Please review the test. After that the way will be open for safe code edits.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Default-Aliases.Tests.ps1:
```diff
@@ -27,7 +27,6 @@ Describe "Verify approved aliases list" -Tags "CI" {
         $CoreWindows = $isCoreCLR -and $IsWindows
         $CoreUnix = $isCoreCLR -and !$IsWindows
-        $FullCLR -or $CoreWindows -or $CoreUnix
         $aliasFullList = @{
```
> "nwsn -> New-PSWorkflowSession"           =             $FullCLR

I don't see this alias on windows powershell ...
```
PS:4> Get-Alias -Definition New-PSWorkflowSession
Get-Alias : This command cannot find a matching alias because an alias with the definition 'New-PSWorkflowSession'
does not exist.
At line:1 char:1
+ Get-Alias -Definition New-PSWorkflowSession
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (New-PSWorkflowSession:String) [Get-Alias], ItemNotFoundException
    + FullyQualifiedErrorId : ItemNotFoundException,Microsoft.PowerShell.Commands.GetAliasCommand
```
Environment
```
Name                           Value
----                           -----
PSVersion                      5.1.14393.1066
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14393.1066
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

iSazonov replied:
😕 I see on 10.0.10240 and on 10.0.16179

daxian-dbw replied:
My bad. It's defined in the module `PSWorkflow` which is not loaded by default. Let's keep it in the list.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Default-Aliases.Tests.ps1:
```diff
@@ -92,7 +91,7 @@ Describe "Verify approved aliases list" -Tags "CI" {
             "gsnp -> Get-PSSnapin"                    =             $FullCLR
-            "gsv -> Get-Service"                      =             $FullCLR -or $CoreWindows -or $CoreUnix
+            "gsv -> Get-Service"                      =             $FullCLR -or $CoreWindows
             "gtz -> Get-TimeZone"                     =                          $CoreWindows
```
`gtz` is also available in windows powershell.

iSazonov replied:
No on 10.0.10240

daxian-dbw replied:
```
PS:3> Get-Alias gtz

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Alias           gtz -> Get-TimeZone                                3.1.0.0    Microsoft.PowerShell.Manag...
```
Available in 10.0.14393.1066
```
Name                           Value
----                           -----
PSVersion                      5.1.14393.1066
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14393.1066
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

iSazonov replied:
What version we should use as a base?

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Default-Aliases.Tests.ps1:
```diff
@@ -198,7 +197,10 @@ Describe "Verify approved aliases list" -Tags "CI" {
-
-        Compare-Object -ReferenceObject $currentAliasList -DifferenceObject $aliaslist | Should Be $null
+        $result = Compare-Object -ReferenceObject $currentAliasList -DifferenceObject $aliaslist
+        # Below 'Should Be' don't show full list wrong aliases so we output them explicitly
```
> don't show full list wrong aliases

doesn't show the list of different aliases

daxian-dbw replied:
This comment hasn't been addressed yet.

iSazonov replied:
We should remove this from `Default-Aliases.Tests.ps1` as the last step in the PR.

daxian-dbw replied:
sounds good.

daxian-dbw commented:
@iSazonov the test looks mostly good. I left a few comments.

iSazonov commented:
@daxian-dbw I fixed the original list of aliases. 
Please could you comment:
1. Currently in `InitialSessionState.cs` all aliases is added with 'ScopedItemOptions.AllScope' attribute and some one still with `ScopedItemOptions.ReadOnly`. I don't see any patterns in that - why not all aliases readonly? Ex.: `Connect-PSSession` - read only but `New-PSSession` - no?
I could agree that some core aliases should be read only but not all.

2. Also from `[Alias()]` an [attribute is added ](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/InitialSessionState.cs#L5886)with `ScopedItemOptions.None`. Why no  'ScopedItemOptions.AllScope'? Why no `ScopedItemOptions.ReadOnly`?

3. There is [exclusion ](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/InitialSessionState.cs#L5302) for `sls` and `Select-String` - where  did it come from?

mirichmo commented:
@daxian-dbw - Please take another look.

daxian-dbw commented:
@iSazonov sorry for coming back so late.
The scope options, yes, thanks for bringing it up --
- `AllScope` indicates the alias is copied to any new scopes that are created
- `ReadOnly` indicates the alias cannot be changed

> I don't see any patterns in that - why not all aliases readonly? Ex.: Connect-PSSession - read only but New-PSSession - no?
I could agree that some core aliases should be read only but not all.

I agree the alias option settings there seem a bit messed up, for example, aliases for `*-PSSession` cmdlets should be consistent in my opinion.
However, some of the aliases are made `AllScope` only for a good reason -- so that user can make them point to customized function/utility as they want. An example [here](https://serverfault.com/questions/452430/how-to-override-the-default-dir-alias-in-powershell) that override the `dir` alias.

> Also from [Alias()] an attribute is added with ScopedItemOptions.None. Why no 'ScopedItemOptions.AllScope'? Why no ScopedItemOptions.ReadOnly?

Yep, `AliasAttribute` now doesn't set any options to an alias, which seems by design to me, because we don't want the module author to decide if their alias definition should be copied to all scopes or can never be changed by the user. For a module cmdlet, a user should be the one to make that decision. PowerShell defines some alias as 'AllScope' or 'ReadOnly' because they are core cmdlets and probably shouldn't be randomly altered by the user.

> There is exclusion for sls and Select-String - where did it come from?

This can be explained by the same `dir` example I gave above. It sounds like very common for people to set `sls` to point to a different function/utility. With `AllScope`, you have to write `Set-Alias ... -Option AllScope` to make it work, which is annoying. I think that's why it was made `None` option.

It turns out there are good reasons to keep alias declarations in InitialSessionState.cs. I suggest we go back to the proposal I made in https://github.com/PowerShell/PowerShell/pull/3595#issuecomment-295951577

mirichmo commented:
@iSazonov - Please update the PR title and description to reflect the current state of the PR. This PR has slowly moved beyond the original intent and I'd like that reflected in the title and description.

iSazonov committed:
Remove unneeded tests

iSazonov commented:
@mirichmo  the PR title and description was updated.

@daxian-dbw Thanks for great comments! I moved aliases to InitialSessionState.cs

iSazonov committed:
Add comment

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Management/commands/management/GetComputerInfoCommand.cs:
```diff
@@ -24,7 +26,6 @@ namespace Microsoft.PowerShell.Commands
     /// </summary>
     [Cmdlet(VerbsCommon.Get, "ComputerInfo",
         HelpUri = "https://go.microsoft.com/fwlink/?LinkId=799466")]
-    [Alias("gin")]
```
I don't think we should move `gin` to InitialSessionState.cs. I think for all aliases that are declared using `Alias()` attribute, we should keep them that way. They are not core cmdlets, and I don't see a problem to have them with 'None' option.

daxian-dbw replied:
For those aliases that are already declared in `InitialSessionState.cs`, I prefer to keep them there for now, for the sake of backward compatibility.
For those aliases that are declared with `AliasAttribute`, I also prefer to keep them that way unless it's a core cmdlet that should be set with options other than 'None'.

iSazonov replied:
I have found _no other_ `[Alias()]` in the repo. So we should keep consistency and use the same approach - place aliases in ` InitialSessionState.cs`.
Also I set `ReadOnly` because later we can replace it on `None` based on feedback but changing  `None` to `ReadOnly` will be breaking change.

daxian-dbw replied:
For those aliases that are already declared in `InitialSessionState.cs`, we should keep them there for backward compatibility. For new aliases, I believe they should be declared using `AliasAttribute`, unless there is an explicit reason to set it with an option other than `None`.

iSazonov replied:
Should we ignore here performance of PowerShell Core start-up?

iSazonov replied:
If we statically add aliases and cmdlets for all core modules we can speed up PowerShell Core start-up.
@daxian-dbw Sorry for perseverance.
@lzybkr Could you please comment too?

daxian-dbw replied:
My opinion doesn't stem from the performance concern. I just think we should no longer keep stuffing unnecessary settings to `InitialSessionState.cs`. Keep alias declaration in individual cmdlet is good for at least 2 reasons:
1. alias declaration is together with the cmdlet definition -- easy to read/discover and the alias exists in the cmdlet as a metadata information which can be analyzed using reflection by tools (if there is any tool doing so).
2. we want to stop the pattern of declaring aliases in a centralized location, so that when multiple people working on different cmdlets at the same time, they will have less conflicts that are resulted from editing the same file.

For those existing alias declarations in `InitialSessionState.cs`, we cannot move them out due to the `ScopeItemOption` settings (BTW, I think for some of the aliases there, the authors probably just copied the existing options without knowing if they should actually be applied). But for new aliases, I think it's better to have them coupled with the cmdlet unless an option other than 'none' is really desired.

lzybkr replied:
There are other benefits to the Alias attribute:
* AllScope is terrible for performance - there is little benefit to have most of these aliases as AllScope
* Consistency - if a non-default runspace is constructed and imports one of these modules, that runspace would not get the aliases defined by default. For example:
```powershell

function DoIt($createArg)
{
    [powershell]::Create($createArg).
        AddScript('$ExecutionContext.InvokeCommand.GetCommand("clc", "Alias")').
        Invoke()
}

DoIt ([System.Management.Automation.RunspaceMode]::NewRunspace)
DoIt ([System.Management.Automation.RunspaceMode]::CurrentRunspace)

$iss = [InitialSessionState]::Create()
$iss.ImportPSModule('Microsoft.PowerShell.Management')
$iss.LanguageMode = 'FullLanguage'

DoIt $iss
```

This prints:
```
CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Alias           clc -> Clear-Content
Alias           clc -> Clear-Content
```

Note - only 2 results instead of 3. This is a little weird because Clear-Content is in `Microsoft.PowerShell.Management`, so why isn't `clc`?

We should perhaps consider adding support for AllScope/ReadOnly in the Alias attribute, perhaps as an internal only option so that it isn't abused.

iSazonov replied:
Thanks for clarify! I totally agree with these now and earlier if we reject performance.

iSazonov replied:
Thanks for clarify!
So a plan is:
1. In the PR make a fix as @daxian-dbw requested
2. Open new Issue (to get feedback for removing AllScope) and then new PR to add _internal_  only option to add support for AllScope/ReadOnly in the Alias attribute and migrate (1) alias and (2) cmdlet declarations from `InitialSessionState.cs` to modules.

iSazonov replied:
@daxian-dbw @lzybkr Thanks for clarify!
 So a plan is:
1.In the PR make a fix as @daxian-dbw requested
2.Open new Issue (to get feedback for removing AllScope) and then new PR to add internal only option to add support for AllScope/ReadOnly in the Alias attribute and migrate (1) alias and (2) cmdlet declarations from  InitialSessionState.cs  to modules.

Is it Ok?

daxian-dbw replied:
Sounds like a plan. When creating the new issue, please make sure all our discussion here are captured in the issue. Thanks @iSazonov!

iSazonov replied:
Done #3796

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Management/commands/management/TimeZoneCommands.cs:
```diff
@@ -14,7 +14,6 @@ namespace Microsoft.PowerShell.Commands
     /// </summary>
     [Cmdlet(VerbsCommon.Get, "TimeZone", DefaultParameterSetName = "Name",
         HelpUri = "https://go.microsoft.com/fwlink/?LinkId=799468")]
-    [Alias("gtz")]
```
Same to `gtz` and `stz` in `TimeZoneCommands.cs`

daxian-dbw commented on src/System.Management.Automation/engine/InitialSessionState.cs:
```diff
@@ -5883,6 +5893,9 @@ internal static class PSSnapInHelpers
                         {
+                            // We add aliases from custom modules with 'ScopedItemOptions.None'
+                            // because only a module consumer (not the module author) decides
+                            // whether the alias is read only and available in all scopes.
```
These comments actually should go to `AliasAttribute`.

iSazonov replied:
Move the comment or copy?

daxian-dbw replied:
My bad. This comment should be put here.
However, the comment may need a bit change, since `AliasAttribute` can be used in built-in module, not custom modules only. Maybe the following is a little better:
"Alias declared by AliasAttribute is set with the option 'ScopedItemOptions.None', because we believe a user of the cmdlet, instead of the author of it, should be the one to decide the option of the alias usage."

iSazonov replied:
Maybe:
"Alias declared by AliasAttribute is set with the option 'ScopedItemOptions.None', because we believe a user of the cmdlet, instead of the author of it, should be the one to decide the option ('ScopedItemOptions.ReadOnly' and/or 'ScopedItemOptions.AllScopes') of the alias usage."

daxian-dbw replied:
Sounds good 👍

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Management/commands/management/GetClipboardCommand.cs:
```diff
@@ -1,4 +1,6 @@
-﻿using System;
+﻿#if !CORECLR
```
`GetClipboardCommand.cs` and `SetClipboardCommand.cs` are currently not included in the build (see [here](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Management/Microsoft.PowerShell.Commands.Management.csproj#L32)). We use those excluded compile entries in `csproj` files as a living document to track what built-in cmdlets are not in powershell core yet. So the `if/def CORECLR` is not necessary in those files.

iSazonov replied:
Fixed.

iSazonov committed:
Remove if/def CORECLR

iSazonov committed:
Move new aliases to modules

iSazonov commented:
If we shall be move alias definitions to _multiple_ files (modules) maybe add tests in `DefaultCommands.Tests.ps1` to check alias's ScopedItemOptions?

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -13,7 +13,6 @@ namespace Microsoft.PowerShell.Commands
     /// Displays the hexidecimal equivalent of the input data.
     /// </summary>
     [Cmdlet(VerbsCommon.Format, "Hex", SupportsShouldProcess = true, HelpUri ="https://go.microsoft.com/fwlink/?LinkId=526919")]
-    [Alias ("fhx")]
```
This alias declaration shouldn't be removed.

iSazonov replied:
Oh, sorry. Fixed.

daxian-dbw commented on src/Modules/Unix/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1:
```diff
@@ -24,7 +24,6 @@ CmdletsToExport= "Format-List", "Format-Custom", "Format-Table", "Format-Wide",
     "Get-Runspace", "Debug-Runspace", "Enable-RunspaceDebug", "Disable-RunspaceDebug",
     "Get-RunspaceDebug", "Wait-Debugger" , "Get-Uptime", "New-TemporaryFile", "Get-Verb", "Format-Hex"
 FunctionsToExport= "Import-PowerShellDataFile"
-AliasesToExport= "fhx"
```
This shouldn't be removed.

iSazonov replied:
Why? Currently `Format-Hex` cmdlet is binary. I believe it was left from the time when the cmdlet was function.
Furthermore we now control all the aliases in the tests and can safely export all aliases by default.

daxian-dbw replied:
If alias is not declared in `.psd1` file, using the alias when the module hasn't been loaded will not trigger the module autoloading. Try the following and you will see it:
```
gmo | rmo
gin
```

iSazonov replied:
Is the behavior different from cmdlets export?

If  we move aliases from InitialSessionState.cs we should add all ones to psd1?
```powershell
gcm | ? Source -eq "Microsoft.PowerShell.Management" | % {Get-Alias -Definition $_.Name -ErrorAction SilentlyContinue}
```

daxian-dbw replied:
> Is the behavior different from cmdlets export?
Not sure what you mean by "cmdlet export"

> If we move aliases from InitialSessionState.cs we should add all ones to psd1?
From what I observed, it seems to be so.

iSazonov replied:
>Not sure what you mean by "cmdlet export".

Defaults is AliasesToExport="*" and CmdletsToExport="*". For first case the module autoload don't works, for second it works?

daxian-dbw replied:
No, it won't work. If you remove a cmdlet from the "CmdletsToExport" then that cmdlet will not be exposed from the module.

iSazonov replied:
Thanks for clarify! It seems I confused with the IntelliSense behavior.

iSazonov replied:
Fixed.

daxian-dbw commented on src/Modules/Windows-Core/Microsoft.PowerShell.Management/Microsoft.PowerShell.Management.psd1:
```diff
@@ -7,7 +7,6 @@ ModuleVersion="3.1.0.0"
 PowerShellVersion="3.0"
 NestedModules="Microsoft.PowerShell.Commands.Management.dll"
 HelpInfoURI = 'https://go.microsoft.com/fwlink/?linkid=390785'
-AliasesToExport = @("gin", "gtz", "stz")
```
This shouldn't be removed.

iSazonov replied:
The same - we now control all the aliases in the tests and can safely export all aliases by default.

daxian-dbw replied:
Same reason as explained above.

iSazonov replied:
Fixed.

daxian-dbw commented on src/Modules/Windows-Core/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1:
```diff
@@ -24,7 +24,6 @@ CmdletsToExport= "Format-List", "Format-Custom", "Format-Table", "Format-Wide",
     "Unblock-File", "Get-Runspace", "Debug-Runspace", "Enable-RunspaceDebug", "Disable-RunspaceDebug",
     "Get-RunspaceDebug", "Wait-Debugger" , "Get-Uptime", "Get-Verb", "Format-Hex"
 FunctionsToExport= "ConvertFrom-SddlString"
-AliasesToExport= "fhx"
```
This shouldn't be removed.

iSazonov replied:
The same - we now control all the aliases in the tests and can safely export all aliases by default.

daxian-dbw replied:
Same reason as explained above.

iSazonov replied:
Fixed.

daxian-dbw commented on src/Modules/Windows-Full/Microsoft.PowerShell.Management/Microsoft.PowerShell.Management.psd1:
```diff
@@ -8,7 +8,6 @@ PowerShellVersion="3.0"
 CLRVersion="4.0"
 NestedModules="Microsoft.PowerShell.Commands.Management.dll"
 HelpInfoURI = 'https://go.microsoft.com/fwlink/?linkid=390785'
-AliasesToExport = @("gcb", "scb", "gin", "gtz", "stz")
```
Shouldn't be removed.

iSazonov replied:
The same - we now control all the aliases in the tests and can safely export all aliases by default.

daxian-dbw replied:
Same reason as explained above.

iSazonov replied:
Fixed.

daxian-dbw commented on src/Modules/Windows-Full/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1:
```diff
@@ -26,7 +26,6 @@ CmdletsToExport= "Format-List", "Format-Custom", "Format-Table", "Format-Wide",
     "Get-Runspace", "Debug-Runspace", "Enable-RunspaceDebug", "Disable-RunspaceDebug", "Get-RunspaceDebug", "Wait-Debugger",
     "ConvertFrom-String", "Convert-String" , "Get-Uptime", "New-TemporaryFile", "Get-Verb", "Format-Hex"
 FunctionsToExport= "ConvertFrom-SddlString"
-AliasesToExport= "CFS", "fhx"
```
shouldn't be removed.

iSazonov replied:
The same - we now control all the aliases in the tests and can safely export all aliases by default.

daxian-dbw replied:
Same reason as explained above.

iSazonov replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/engine/InitialSessionState.cs:
```diff
@@ -5016,6 +5016,8 @@ internal static SessionStateAliasEntry[] BuiltInAliases
                     new SessionStateAliasEntry("epcsv",
                         "Export-Csv",      "", ScopedItemOptions.ReadOnly | ScopedItemOptions.AllScope),
+                    new SessionStateAliasEntry("fhx",
+                        "Format-Hex",      "", ScopedItemOptions.ReadOnly | ScopedItemOptions.AllScope),
```
Shouldn't be added here.

iSazonov replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/engine/InitialSessionState.cs:
```diff
@@ -5279,6 +5281,8 @@ internal static SessionStateAliasEntry[] BuiltInAliases
                         "Wait-Job",  "", ScopedItemOptions.AllScope),
 #if !CORECLR
+                    new SessionStateAliasEntry("cfs",
+                        "ConvertFrom-String", "", ScopedItemOptions.AllScope),
```
shouldn't be added here. `ConvertFrom-String` uses `AliasAttribute` to declare the alias.
```
PS:3> $t = gcm  ConvertFrom-String | % ImplementingType
PS:4> $t.GetCustomAttributes($false) | ? { $_.GetType().Name -eq "AliasAttribute" }

AliasNames TypeId
---------- ------
{CFS}      System.Management.Automation.AliasAttribute
```

iSazonov replied:
Removed.

daxian-dbw commented on src/System.Management.Automation/engine/InitialSessionState.cs:
```diff
@@ -5883,6 +5887,10 @@ internal static class PSSnapInHelpers
+                            // Alias declared by AliasAttribute is set with the option 'ScopedItemOptions.None',
+                            // because we believe a user of the cmdlet, instead of the author of it,
+                            // should be the one to decide the option
+                            // ('ScopedItemOptions.ReadOnly' and/or 'ScopedItemOptions.AllScopes') of the alias usage."
```
This comment probably should be removed now. When making the change to add internal support to `AliasAttribute` to allow specifying options, we can have a more accurate comment here.

iSazonov replied:
Today the comment is correct. After improving the AliasAttribute the code here will be changed too and we could adjust the comment in that time.

daxian-dbw replied:
That's reasonable #closed

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Clear-Variable.Tests.ps1:
```diff
@@ -114,12 +114,6 @@ Describe "Clear-Variable" -Tags "CI" {
-	clv -Name var1
-	$var1 | Should BeNullOrEmpty
-	{ Get-Variable var1 } | Should Not Throw
-		}
```
This test is not checking if the alias exists, but actually testing the functionality of `clear-variable`, so it shouldn't be removed.

iSazonov replied:
Previous test do that so the test is alias related only.

daxian-dbw replied:
#Closed.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Invoke-Expression.Tests.ps1:
```diff
@@ -6,25 +6,13 @@ Describe "Invoke-Expression" -Tags "CI" {
 
-	It "Should return the echoed text for the alias" {
-	    (iex -command "echo pestertest2") | Should Be "pestertest2"
-	}
```
Same here. It's not testing whether the alias `iex` eixsts, but testing the functionality of `iex`. So it shouldn't be removed.

iSazonov replied:
Previous test do that so the test is alias related only.

daxian-dbw replied:
#Closed

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Invoke-Expression.Tests.ps1:
```diff
@@ -6,25 +6,13 @@ Describe "Invoke-Expression" -Tags "CI" {
-	    $testcommand = "echo pestertestscript"
-	    $testcommand | Add-Content -Path "$testfile"
-	    (iex "& '$testfile'") | Should Be "pestertestscript"
-	    Remove-Item "$testfile"
```
Again, this is testing the functionality, not the existence of alias.

iSazonov replied:
Previous test do that so the test is alias related only.

daxian-dbw replied:
#Closed

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/New-Variable.Tests.ps1:
```diff
@@ -84,13 +84,6 @@ Describe "New-Variable" -Tags "CI" {
-	$var1 | Should BeNullOrEmpty
-	nv var2 2
-	$var2 | Should Be 2
-    }
```
same here. Shouldn't be removed.

iSazonov replied:
Previous test do that so the test is alias related only.

daxian-dbw replied:
#Closed

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Remove-PSBreakpoint.Tests.ps1:
```diff
@@ -49,16 +49,6 @@ Get-Date
-	    $(Get-PSBreakpoint).Id.length | Should Be ($NumberOfBreakpoints -1)
-	}
-    }
-
```
Same here. Shouldn't be removed.

iSazonov replied:
Previous test do that so the test is alias related only.

daxian-dbw replied:
#Closed

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Remove-Variable.Tests.ps1:
```diff
@@ -28,17 +28,6 @@ Describe "Remove-Variable" -Tags "CI" {
-
-	$var1 | Should Be #nothing.  it should be Nothing at all.
-
-    }
```
same here. Shouldn't be removed.

iSazonov replied:
Previous test do that so the test is alias related only.

daxian-dbw replied:
#Closed

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -9,14 +9,10 @@ Describe "Select-String" -Tags "CI" {
-    it "Should return an array data type when multiple matches are found" {
-        $result = $testinputtwo | Select-String -Pattern "hello"
-        ,$result | Should BeOfType "System.Array"
-    }
```
Same here. Both tests shouldn't be removed.

iSazonov replied:
The first - Previous test do that so the test is alias related only.
The second is not removed - it is below - diff trick 😕

daxian-dbw replied:
#Closed

daxian-dbw commented on test/powershell/engine/DefaultCommands.Tests.ps1:
```diff
@@ -0,0 +1,497 @@
+            # instead of Get-Alias and Get-Command
+            # but latest is more simple use.
+            #$iis = [initialsessionstate]::CreateDefault()
+            #$currentCmdletList = $iis.Commands | Where-Object { $_.CommandType -eq "Cmdlet"} | Select-Object -ExpandProperty Name
```
These comments don't seem necessary to keep.

iSazonov replied:
Yes. Removed.

daxian-dbw commented on test/powershell/engine/DefaultCommands.Tests.ps1:
```diff
@@ -0,0 +1,497 @@
+        # if all aliases is Ok we output nothing
+        $result | Write-Host
+        $result | Should Be $null
+    }
```
Isn't this test the same as the one in `Default-Aliases.Tests.ps1`? Are you planning to remove `Default-Aliases.Tests.ps1` and replace it with this file?

iSazonov replied:
Yes, I'll remove the duplication. I leave it for easy review and for checking new tests in DefaultCommands.Tests.ps1.

daxian-dbw commented:
> If we shall move alias definitions to multiple files (modules) maybe add tests in DefaultCommands.Tests.ps1 to check alias's ScopedItemOptions?

That sounds great.

iSazonov committed:
Fix [Alias ("fhx")]

iSazonov committed:
Remove 'cfs' alias

iSazonov committed:
Remove comment

iSazonov committed:
Remove unneeded alias test

iSazonov committed:
Fix changes in  .psd1 files

iSazonov committed:
Add checks for alias options and minor refactoring

daxian-dbw commented:
@iSazonov The changes look good to me, and special thanks for the new tests covering the complete list of built-in aliases/cmdlets!
I think it's good to go as soon as you remove the old `Default-Aliases.Tests.ps1`.

iSazonov committed:
Remove duplicated alias tests

iSazonov commented:
@daxian-dbw I removed `Default-Aliases.Tests.ps1` and move test for `more` to `DefaultCommands.Tests.ps1`

iSazonov commented:
@daxian-dbw Thanks for help and great comments!
